### PR TITLE
[Stack 821] -Convert Session Persistence to TaskFlow

### PR DESCRIPTION
**Pull request by omkartelee01**
closes [STACK-812](https://a10networks.atlassian.net/browse/STACK-812)

### CODE REVIEW

File: a10_octavia/controller/worker/tasks/handler_l7rule.py [hunk 1]
```diff
@@ -26,55 +26,65 @@
 class L7RuleParent(object):
 
     def set(self, l7rule, listeners, vthunder):
-        """ Execute create l7rule """
+        """Execute create l7rule """
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/handler_l7rule.py]: Remove the space at the end

File: a10_octavia/controller/worker/tasks/handler_l7rule.py [hunk 2]
```diff
@@ -26,55 +26,65 @@
...
+            LOG.debug("Listener updated successfully: %s", listener.id)
+        except Exception as e:
+            LOG.exception("Failed to create/update l7rule: %s", str(e))
+            raise
 
 
 class CreateL7Rule(L7RuleParent, BaseVThunderTask):
-    """ Task to create L7Rule """
+
+    """Task to create L7Rule """
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/handler_l7rule.py]: Remove these spaces

File: a10_octavia/controller/worker/tasks/handler_l7rule.py [hunk 3]
```diff
@@ -26,55 +26,65 @@
...
 
     def execute(self, l7rule, listeners, vthunder):
         self.client_factory(vthunder)
         self.set(l7rule, listeners, vthunder)
 
 
 class UpdateL7Rule(L7RuleParent, BaseVThunderTask):
-    """ Task to update L7Rule """
+
+    """Task to update L7Rule """
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/handler_l7rule.py]: Remove the space

File: a10_octavia/controller/worker/tasks/handler_l7rule.py [hunk 4]
```diff
@@ -83,53 +93,59 @@ def execute(self, l7rule, listeners, vthunder, update_dict):
 
 
 class DeleteL7Rule(BaseVThunderTask):
-    """ Task to delete a L7rule and disassociate from provided pool """
+
+    """Task to delete a L7rule and disassociate from provided pool """
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/handler_l7rule.py]: Remove the space

File: a10_octavia/controller/worker/tasks/handler_l7rule.py [hunk 5]
```diff
@@ -83,53 +93,59 @@ def execute(self, l7rule, listeners, vthunder, update_dict):
...
 
     def execute(self, l7rule, listeners, vthunder):
-        """ Execute delete L7Rule """
+        """Execute delete L7Rule """
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/handler_l7rule.py]: Remove the space

File: a10_octavia/controller/worker/tasks/persist_tasks.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+LOG = logging.getLogger(__name__)
+
+
+class HandleSessionPersistenceDelta(BaseVThunderTask):
+
+    def execute(self, vthunder, pool):
+        c_pers, s_pers, sp = None, None, None
+        sp = pool.session_persistence
+        c_pers, s_pers = utils.get_sess_pers_templates(pool)
+        import rpdb; rpdb.set_trace()
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/persist_tasks.py]: Remove this...

File: a10_octavia/controller/worker/tasks/persist_tasks.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+LOG = logging.getLogger(__name__)
+
+
+class HandleSessionPersistenceDelta(BaseVThunderTask):
+
+    def execute(self, vthunder, pool):
+        c_pers, s_pers, sp = None, None, None
+        sp = pool.session_persistence
+        c_pers, s_pers = utils.get_sess_pers_templates(pool)
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/persist_tasks.py]: We don't need these as they aren't used at all later on.

File: a10_octavia/controller/worker/tasks/persist_tasks.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+from a10_octavia.controller.worker.tasks import utils
+from a10_octavia.common.a10constants import SP_OBJ_DICT, PERS_TYPE
+
+LOG = logging.getLogger(__name__)
+
+
+class HandleSessionPersistenceDelta(BaseVThunderTask):
+
+    def execute(self, vthunder, pool):
+        c_pers, s_pers, sp = None, None, None
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/persist_tasks.py]: Remove this for 2 reason.
1. (Future ref): No need to define these as None like this if we are going to assign them right away. 
2. We don't use c_pers or s_pers

File: a10_octavia/controller/worker/tasks/persist_tasks.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        sp = pool.session_persistence
+        c_pers, s_pers = utils.get_sess_pers_templates(pool)
+        import rpdb; rpdb.set_trace()
+        if sp and sp.type and sp.type in SP_OBJ_DICT:
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/persist_tasks.py]: This should be:
```python
if pool.session_persistence in SP_OBJ_DICT:
```

File: a10_octavia/controller/worker/tasks/persist_tasks.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+                    sp_template = getattr(axapi_client.slb.template, sp_type)
+                    sp_template.delete(pool.id)
+                except acos_errors.NotFound:
+                    pass
+
+            sp_template = getattr(
+                axapi_client.slb.template, SP_OBJ_DICT[sp.type])
+
+            try:
+                if sp.cookie_name:
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/persist_tasks.py]: This becomes:
```python
if pool.session_persistence.cookie_name:
```

File: a10_octavia/controller/worker/tasks/persist_tasks.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+                pass
+            except Exception:
+                LOG.exception(
+                    "Failed to create session persistence for pool: %s", pool.id)
+                raise
+
+    def revert(self, vthunder, pool, *args, **kwargs):
+        axapi_client = self.client_factory(vthunder)
+        sp_template = getattr(
+            axapi_client.slb.template, SP_OBJ_DICT[sp.type])
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/persist_tasks.py]: `sp` is not defined here. Instead use `pool.session_persistence`

File: a10_octavia/tests/common/a10constants.py [hunk 12]
```diff
@@ -16,3 +16,4 @@
 MOCK_POOL_ID = 'mock-pool-1'
 MOCK_MEMBER_ID = 'mock-member-1'
 MOCK_LOAD_BALANCER_ID = 'mock-lb-1'
+MOCK_COOKIE_NAME = 'cookie1'
```
Comment (hthompson6) [inline a10_octavia/tests/common/a10constants.py]: Follow the other naming format.

```python
MOCK_COOKIE_NAME = 'mock-cookie-1'
```

File: a10_octavia/tests/unit/base.py [hunk 13]
```diff
@@ -15,6 +15,7 @@
 
 import mock
 from unittest.mock import patch
+import acos_client.errors as acos_errors
```
Comment (hthompson6) [inline a10_octavia/tests/unit/base.py]: Unused import

File: a10_octavia/controller/worker/tasks/utils.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
+#    Copyright 2019, A10 Networks
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/utils.py]: Copyright 2020

File: a10_octavia/controller/worker/tasks/persist_tasks.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
+#    Copyright 2019, A10 Networks
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/persist_tasks.py]: Copyright 2020

File: a10_octavia/controller/worker/tasks/handler_l7rule.py (hunk 1 above)
Comment (omkartelee01) [reply a10_octavia/controller/worker/tasks/handler_l7rule.py]: Removing this line, as set is not expected to have this line, it should be only in CreateL7RuleTask

File: a10_octavia/controller/worker/tasks/handler_l7rule.py (hunk 2 above)
Comment (omkartelee01) [reply a10_octavia/controller/worker/tasks/handler_l7rule.py]: corrected

File: a10_octavia/controller/worker/tasks/handler_l7rule.py (hunk 4 above)
Comment (omkartelee01) [reply a10_octavia/controller/worker/tasks/handler_l7rule.py]: corrected, removed extra space

File: a10_octavia/controller/worker/tasks/handler_l7rule.py (hunk 5 above)
Comment (omkartelee01) [reply a10_octavia/controller/worker/tasks/handler_l7rule.py]: Removed this redundant line, as class is already defining exact same thing

File: a10_octavia/controller/worker/tasks/persist_tasks.py (hunk 6 above)
Comment (omkartelee01) [reply a10_octavia/controller/worker/tasks/persist_tasks.py]: I accidentally overwritten the old file that was in my dev machine which was not cleaned up, my bad!

File: a10_octavia/controller/worker/tasks/persist_tasks.py (hunk 8 above)
Comment (omkartelee01) [reply a10_octavia/controller/worker/tasks/persist_tasks.py]: I accidentally overwritten the old persist_task.py that was in my dev machine which was not cleaned up, and it went to PR, my bad!

File: a10_octavia/controller/worker/tasks/persist_tasks.py (hunk 10 above)
Comment (omkartelee01) [reply a10_octavia/controller/worker/tasks/persist_tasks.py]: resolved

File: a10_octavia/controller/worker/tasks/handler_l7rule.py [hunk 16]
```diff
@@ -26,110 +26,121 @@
...
-            LOG.error(str(e))
-            LOG.info("Error occurred")
+            LOG.exception("Failed to create/update l7rule: %s", str(e))
+            raise
 
 
 class CreateL7Rule(L7RuleParent, BaseVThunderTask):
-    """ Task to create L7Rule """
 
+    """Task to create L7Rule"""
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/handler_l7rule.py]: Not sure why these are positioned right over the execute function. This is a class level comment.

```python
class ClassName(...):
    """Comment goes here"""

    def func(...):
        pass
```

File: a10_octavia/controller/worker/tasks/handler_l7rule.py [hunk 17]
```diff
@@ -26,110 +26,121 @@
...
     def execute(self, l7rule, listeners, vthunder):
         self.client_factory(vthunder)
         self.set(l7rule, listeners, vthunder)
 
 
 class UpdateL7Rule(L7RuleParent, BaseVThunderTask):
-    """ Task to update L7Rule """
 
+    """Task to update L7Rule"""
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/handler_l7rule.py]: Not sure why these are positioned right over the execute function. This is a class level comment.

```python
class ClassName(...):
    """Comment goes here"""

    def func(...):
        pass
```

File: a10_octavia/controller/worker/tasks/handler_l7rule.py [hunk 18]
```diff
@@ -26,110 +26,121 @@
...
     def execute(self, l7rule, listeners, vthunder, update_dict):
         l7rule.__dict__.update(update_dict)
         self.client_factory(vthunder)
         self.set(l7rule, listeners, vthunder)
 
 
 class DeleteL7Rule(BaseVThunderTask):
-    """ Task to delete a L7rule and disassociate from provided pool """
 
+    """Task to delete a L7rule and disassociate from provided pool"""
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/handler_l7rule.py]: Not sure why these are positioned right over the execute function. This is a class level comment.

```python
class ClassName(...):
    """Comment goes here"""

    def func(...):
        pass
```

File: a10_octavia/controller/worker/tasks/persist_tasks.py [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+import acos_client.errors as acos_errors
+
+from a10_octavia.controller.worker.tasks.common import BaseVThunderTask
+from a10_octavia.controller.worker.tasks import utils
+from a10_octavia.common.a10constants import SP_OBJ_DICT, PERS_TYPE
+
+LOG = logging.getLogger(__name__)
+
+
+class HandleSessionPersistenceDelta(BaseVThunderTask):
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/persist_tasks.py]: Let's add a class level comment here as is done with the other tasks classes.

**Pull request merged**